### PR TITLE
fix(lsp): convert_input_to_markdown_lines: preserve plaintext

### LINKED ISSUE #14746: language server client does not respect markupkind plaintext
- `nvim --version`: nvim --version
NVIM v0.5.0-dev+1384-gf2906a466-dirty
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/dch/neovim/build/config -I/home/dch/neovim/src -I/home/dch/neovim/.deps/usr/include -I/usr/include -I/home/dch/neovim/build/src/nvim/auto -I/home/dch/neovim/build/include
Compiled by dch@DESKTOP-NA4O3IV

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/dch/.local/nvim/share/nvim"

Run :checkhealth for more info
- Operating system/version: Ubunto 20.04
- Terminal name/version: wsltty
- `$TERM`: tmux-256color

[ ] `vim -u DEFAULTS` (version: ) behaves differently

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
# Alternative for shell-related problems:
# env -i TERM=ansi-256color "$(which nvim)"

```

### Actual behaviour

Hover text is presented as markdown even though the server declared that it was plaintext.

### Expected behaviour

plaintext should be presented as plain text.

I am using a python language server.  Python conventions
- do not write comments in markdown
- do make heavy use of `snake_case` and `__dunder__` methods and suchlike

The server does not attempt to escape everything and so I have configured it to return hover etc with kind `plaintext`.

But the neovim language client presents it as markdown anyway.  This is annoying because eg the markdown syntax insists that an underscore in the middle of a word (method name) is an error, and tries to present `__foo__` as __foo__ , and so on.

cf https://github.com/pappasam/jedi-language-server/issues/144#issuecomment-856445819

### PR DESCRIPTION
**Pull request by folke**
Text coming from an lsp server can have markup kind `markdown` or `plaintext`. Right now we always assume it's `markdown`.

This PR properly handles `plaintext` within `convert_input_to_markdown_lines`, by wrapping it in `<text>...</text>`

Just passing the contents to `open_floating_preview` with syntax `plaintex` wouldn't work in other places like `signature_help`, where we build up a `markdown` document containing the markup from `documentation` and params.

By making sure `convert_input_to_markdown_lines` properly deals with `plaintext`, we can still use `markdown` internally for the rest of the text, but now the `plaintext` sections will be properly stylzed (or not styled) with syntax `plaintex`.

This PR also fixes:
* proper matching of beginning / end parts of a code block, `<pre>`, `<code>`, `<text>`. Previously, something like `<pre>```</pre>` would have ended the code block when encountering the backticks, instead of extending the block until the correct `</pre>` ending
* added `<code>` blocks that might appear in comments (works the same as `<pre>`)
* allow any number of whitespace preceding and following a beginning or end marker of a code block.

Fixes #14746

[LSP Spec for markupContent](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#markupContent)

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -806,14 +806,22 @@ function M.convert_input_to_markdown_lines(input, contents)
...
-      -- should make sure we handle this correctly.
+      -- The kind can be either plaintext or markdown.
+      -- If it's plaintext, then wrap it in a <text></text> block
 
       -- Some servers send input.value as empty, so let's ignore this :(
+      input.value = input.value or ''
+
+      if input.kind == "plaintext" then
+        -- very unlikely, but but escape the end tag, just to be sure
+        input.value = input.value:gsub("</text>", "\\</text>")
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:817]: Do we need this workaround? I feel like we should instead report this upstream in the language server than having a workaround for something which may/may not occur.

**Review by mjlbach — APPROVED** [review]

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (folke) [reply runtime/lua/vim/lsp/util.lua:817]: You mean the escaping?

I can leave it out. Very unlikely it will ever occur. Happy to change it if you want.

Comment (folke) [reply runtime/lua/vim/lsp/util.lua:817]: Not at home right now, but I'll remove it. Not really needed.

**Review by mfussenegger — APPROVED** [review]

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:817]: kk will merge after

Comment (folke) [reply runtime/lua/vim/lsp/util.lua]: done

**Comment by tomtomjhj** [comment]
Is `"plaintex"` the intended filetype for plain text?  I thought it was typo since plaintex is for the plain variant of TeX.

**Comment by jamessan** [comment]
> Is `"plaintex"` the intended filetype for plain text? I thought it was typo since plaintex is for the plain variant of TeX.

Nice catch.  Both `"plaintex"` and `"plaintext"` are used in the PR.  The comments make it sound like we may get `"plaintext"` from a server and then we convert that to a filetype internally.  The latter should be `"text"`.

Do you want to open a PR?

**Pull request merged**